### PR TITLE
Endless Iterators

### PR DESCRIPTION
**Pull request by fee1-dead**
Endless Iterators: iterators that are either empty or yields elements infinitely.

Implement `ExactSizeIterator` for `Take` when the underlying iterator 
`: ExactSizeIterator | EndlessIterator`. This could have benefits since `repeat(item).take(x)` is pretty common.

### CODE REVIEW

**Review by the8472 — COMMENTED** [review]
I know it's a draft, so feel free to ignore comments that you're already aware of.

File: library/core/src/iter/traits/marker.rs [hunk 1]
```diff
@@ -72,3 +72,26 @@ pub unsafe trait InPlaceIterable: Iterator {}
 #[unstable(feature = "trusted_step", issue = "85731")]
 #[rustc_specialization_trait]
 pub unsafe trait TrustedStep: Step {}
+
+/// An iterator that is either empty or endless.
+///
+/// The iterator reports a size hint that is either `0, Some(0)` or `usize::MAX, None`.
```
[the remark this answers is not in the document]
Comment (scottmcm) [reply library/core/src/iter/traits/marker.rs:78]: You had be worried for a second here -- I think you mean `0, None` even when it's not empty? https://doc.rust-lang.org/1.56.0/src/core/iter/adapters/flatten.rs.html#313-332

If `Flatten` can return `0, Some(0)` when it's not empty, please file a bug!

**Comment by scottmcm** [comment]
TBH there are a few places where `InfiniteIterator` would be handy -- especially since it could have a default method, like
```rust
pub trait InfiniteIterator: Iterator {
    fn next_infinite(&mut self) -> Self::Item {
        match self.next() {
            Some(x) => x,
            None => infinite_iterator_was_finite(),
        }
    }
}

#[cold]
fn infinite_iterator_was_finite() -> ! {
    unreachable!()
}
```

File: library/core/src/iter/traits/marker.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (the8472) [reply library/core/src/iter/traits/marker.rs:78]: I meant `(0, None)`

**Comment by the8472** [comment]
There are many more optimizations we could apply to iterators, but the problem is that they all require a swarm of unsafe marker traits.

* infinite
* size guaranteed fit into `usize` (but not necessarily an exact size).
  It could be factored out from from `TrustedRandomAccess` so it can also be combined with `TrustedLen` instead (the combination would give us `TrustedExactLength`). Together with the infinite case it would make `Zip` specializations more powerful.
* sorted (consistent with `Ord`) useful for btree and binary heap, this could allow `min()` or `max()` to shortcircuit even after `skip()` or `filter()`
* `clone` does not mess with the size (so we can optimize `Cycle`)

And a bunch of other things I have thought of before but can't recall atm.


I would submit some of these myself but they are very niche (at least for the initial optimizations that I can think of) and it already takes a long time to get more implementations for existing traits in (e.g. #85528). Even if the libs mind is willing the reviewer flesh is weak 😅.

**Comment by scottmcm** [comment]
> There are many more optimizations we could apply to iterators, but the problem is that they all require a swarm of unsafe marker traits.

Yes, but that's why I mentioned infinite as a *safe* trait -- not just something that could be used internally for magic.

**Pull request closed without merging**